onnxscript/evaluator.py

```diff
@@ -0,0 +1,310 @@
+            return len(scan_body.output)
+        if schema.name == "Loop":
+            loop_body = kwargs["body"]
+            return len(loop_body.output) - 1
+    return len(schema.outputs)
+
+
+_cache_models = {}
+
+
+def _cache_(model, providers):
+    serialized = model.SerializeToString()
+    key = serialized, tuple(providers)
+    if key in _cache_models:
+        return _cache_models[key]
+    sess = InferenceSession(serialized, providers=providers)
+    _cache_models[key] = sess
+    return sess
+
+
+def os_to_ort_value(v):
+    """Converts an onnxscript encoding of an ONNX value into the encoding used by ORT."""
+    if isinstance(v, tensor.Tensor):
+        return v.value
+    if isinstance(v, list):
+        return v
+    if v is None:
+        # Treated as a static-optional value.
+        # Dynamic optional None not yet supported.
+        return v
+    if isinstance(v, np.ndarray):
+        return v
+    raise TypeError(f"Unexpected ORT value type {type(v)}.")
+
+
+def ort_to_os_value(v):
+    """Converts an ORT encoding of an ONNX value into the encoding used by onnxscript."""
+    if isinstance(v, np.ndarray):
+        return tensor.Tensor(v)
+    if isinstance(v, list):
+        return v
+    if v is None:
+        raise TypeError("Dynamic optional values not yet supported.")
+    raise TypeError(f"Unexpected ORT value type {type(v)}.")
+
+
+def call_ort(schema, args, kwargs, implicit_args=None):
+    implicit_args = implicit_args or {}
+    # Convert input values to ORT representation-type:
+    args = [os_to_ort_value(x) for x in args]
+    implicit_args = {k: os_to_ort_value(v) for k, v in implicit_args.items()}
+
+    # Construct ONNX model with a single op call:
+    inputs = [_rename_io("input", i, arg) for i, arg in enumerate(args)]
+
+    num_outputs = compute_num_outputs(schema, *args, **kwargs)
+    outputs = [f"output{str(i)}" for i in range(num_outputs)]
+
+    node = onnx.helper.make_node(schema.name, inputs, outputs, domain=schema.domain, **kwargs)
+    input_value_infos = utils.values_to_value_infos(zip(inputs, args))
+    implicit_value_infos = utils.values_to_value_infos(implicit_args.items())
+    output_value_infos = [onnx.helper.make_value_info(name, TypeProto()) for name in outputs]
+
+    graph = onnx.helper.make_graph(
+        [node], "node_graph", input_value_infos + implicit_value_infos, output_value_infos
+    )
+    opset_id = onnx.helper.make_opsetid(schema.domain, schema.since_version)
+    model = onnx.helper.make_model(
+        graph,
+        opset_imports=[opset_id],
+        ir_version=irbuilder.select_ir_version(schema.since_version, domain=schema.domain),
+    )
+    model = onnx.shape_inference.infer_shapes(model)
+    # onnx.checker.check_model(model)
+    try:
+        sess = _cache_(model, ["CPUExecutionProvider"])
+    except (Fail, InvalidGraph, InvalidArgument) as e:
+        raise RuntimeError(
+            f"Unable to create onnxruntime InferenceSession "
```

Comment:
You should recall the schema.name in the error message. I guess that's what user would know, not the ONNX graph.

Reply:
Done, thanks! (This is all pre-existing code, not part of current PR ... but I agree with your comment.)